Implement upload portal for missouri and mississippi court opinions

quevon24 opened the pull request:
These are the implementations for the upload portal for mo, moctapp, miss and missctapp.

Based on what I saw on their sites, i define the following fields for the forms: caption, docket number, publication date, optional lead author and pdf file, only for Mississippi a summary field is required. I didn't find that they had neutral citations like maine or tennessee.

The information I added about the courts was extracted from their pages

![image](https://github.com/freelawproject/courtlistener/assets/6474994/06d8bac7-2153-42df-bcc4-019574b9ac26)

![image](https://github.com/freelawproject/courtlistener/assets/6474994/860ad2a1-7944-4b17-9459-de10d6827113)

![image](https://github.com/freelawproject/courtlistener/assets/6474994/4f7587ab-cc82-4830-ab4e-9509c70f7ecb)

![image](https://github.com/freelawproject/courtlistener/assets/6474994/18dfb4fd-3c6e-44fe-a964-7c3fd6a92669)

I added a back button in the form and an asterisk to indicate required fields.

![image](https://github.com/freelawproject/courtlistener/assets/6474994/46810f02-3bfe-4ab0-9b75-4beeda29adc5)



Tennessee and Maine maintain the same functionality, they just use the same base form.

**miss fields:**
- Caption
- Docket Number
- Publication Date
- Lead Author
- Author string
- Judges string
- Disposition
- Summary
- Download url
- Opinion pdf

**missctapp fields:**
- Caption
- Docket Number
- Publication Date
- Lead Author
- Author string
- Judges string
- Disposition
- Summary
- Download url
- Opinion pdf

**mo fields:**
- Caption
- Docket Number
- Publication Date
- Lead Author
- Author string
- Judges string
- Disposition
- Download url
- Opinion pdf

**moctapp fields:**
- Caption
- Docket Number
- Publication Date
- Lead Author
- Author string
- Judges string
- Disposition
- Download url
- Opinion pdf

**Steps:**

Add the new courts to courtlistener: 
- moctapped
- moctappsd
- moctappwd

Add new user groups and assign users to the groups: 

- uploaders_mo
- uploaders_moctapp
- uploaders_moctapped
- uploaders_moctappsd
- uploaders_moctappwd
- uploaders_miss
- uploaders_missctapp

flooie commented on cl/opinion_page/templates/tn-court.html:
```diff
@@ -3,152 +3,182 @@
-      </div>
-    </div>
-  </div>
+			</div>
+			<div class="col-xs-4">
+				<h3>Neutral Citations</h3>
+				<p>Beginning on January 1st, 2020 the Tennessee Court of Workers’
+					Compensation Claims and the Appeals Board will accept pleadings,
+					briefs, and other filings that include <a
+							href="{% url "faq" %}#explain-neutral-citations">vendor
+						neutral citations</a> to prior decisions as discussed below. All
+					previous decisions have been retroactively assigned such a citation.
+					These citations are created for each decision by the court itself
+					instead of by a third party publisher. As a result, they can be
+					cited immediately when the decision is published by the court.
```
can we fix this whitespace-  im not sure if this is even touching tn

flooie commented on cl/opinion_page/templates/court.html:
```diff
@@ -3,100 +3,119 @@
-      <div class="col-xs-8">
-        <h1>Decisions of the {{ court }}
-        </h1>
-        <div class="lead gray">
-          <p>The Supreme Judicial Court of Maine has partnered with the non-profit organization, Free Law Project, to host its court decisions on CourtListener.com. Free Law Project is proud to support this partnership and makes the decisions available online as they are released by the court.</p>
-        </div>
-      </div>
-    </div>
+	<div id="court_homepage" class="col-xs-12">
+		<div class="row">
+			<div class="col-xs-4">
+				<img src="{% static 'img/' %}{{ pk }}-court.jpg" class="img-responsive"
+					 width="460" height="206">
+			</div>
+			<div class="col-xs-8">
+				<h1>Decisions of the {{ court }}
+				</h1>
+				<div class="lead gray">
+					<p>The Supreme Judicial Court of Maine has partnered with the
+						non-profit organization, Free Law Project, to host its court
+						decisions on CourtListener.com. Free Law Project is proud to
+						support this partnership and makes the decisions available
+						online as they are released by the court.</p>
```
this also just looks like white space changes -can we please not add whitespace where its not needed

flooie commented on cl/opinion_page/forms.py:
```diff
@@ -237,6 +258,70 @@ class CourtUploadForm(forms.Form):
+        ),
+    )
+
+    download_url = forms.URLField(
+        validators=[URLValidator()],
+        label="Download Url",
+        required=False,
+        widget=forms.TextInput(
+            attrs={
+                "class": "form-control",
+                "placeholder": "The URL where the document is originally located",
+                "autocomplete": "off",
+            }
+        ),
+    )
```
Is this the correct description?  since we aren't using a scraper and the upload is being done manually, perhaps we dont bother?  @mlissner

mlissner replied:
The description seems right to me. It seems like a nice thing to have the URL if we can.

flooie commented:
@quevon24 I think we talked about using seal-rookery to generate the seals.

flooie commented on cl/opinion_page/views.py:
```diff
@@ -157,9 +196,33 @@ async def court_publish_page(request: HttpRequest, pk: int) -> HttpResponse:
             )
 
-    form = await sync_to_async(CourtUploadForm)(pk=pk)
+    # Fix mypy errors
+    upload_form: Any
+
+    if pk == "tennworkcompcl":
+        upload_form = TennWorkCompClUploadForm
+    elif pk == "tennworkcompapp":
+        upload_form = TennWorkCompAppUploadForm
+    elif pk == "me":
+        upload_form = MeCourtUploadForm
+    elif pk == "mo":
+        upload_form = MoCourtUploadForm
+    elif pk == "moctapp":
+        upload_form = MoCtAppCourtUploadForm
+    elif pk == "moctapped":
+        upload_form = MoCtAppEdCourtUploadForm
+    elif pk == "moctappsd":
+        upload_form = MoCtAppSdCourtUploadForm
+    elif pk == "moctappwd":
+        upload_form = MoCtAppWdCourtUploadForm
+    elif pk == "miss":
+        upload_form = MissCourtUploadForm
+    else:
+        upload_form = MissCtAppCourtUploadForm
+
+    form = await sync_to_async(upload_form)(pk=pk)
```
Cant we do something like this 
```
upload_form_classes = {
    "tennworkcompcl": TennWorkCompClUploadForm,
    "tennworkcompapp": TennWorkCompAppUploadForm,
    "me": MeCourtUploadForm,
    "mo": MoCourtUploadForm,
    "moctapp": MoCtAppCourtUploadForm,
    "moctapped": MoCtAppEdCourtUploadForm,
    "moctappsd": MoCtAppSdCourtUploadForm,
    "moctappwd": MoCtAppWdCourtUploadForm,
    "miss": MissCourtUploadForm
}
upload_form = upload_form_classes.get(pk, None)
```

flooie replied:
it seems a lot nicer than that long list.

flooie commented on cl/opinion_page/views.py:
```diff
@@ -176,7 +239,14 @@ async def court_publish_page(request: HttpRequest, pk: int) -> HttpResponse:
+        request,
+        "publish.html",
+        {
+            "court_image": f"img/{pk}.jpg",
```
maybe we cant use seal rookery because of CORS - but wouldnt that be nicer than storing all these images

mlissner replied:
If CORS is the problem, we can add seals.free.law to our allowed list in our settings and fix that.

flooie replied:
great.  thanks.  Currently its a problem as we actually use seals on the Maine uploader page but its blocked from loading

quevon24 replied:
The problem was with the CSP, you just need to add the domain seal rockery to CSP_IMG_SRC in security.py

flooie commented on cl/opinion_page/forms.py:
```diff
@@ -237,6 +258,70 @@ class CourtUploadForm(forms.Form):
+        widget=forms.Textarea(attrs={"class": "form-control"}),
+    )
+
+    # TODO this could be a select field with hardcoded options for a specific court
```
?

quevon24 replied:
I thought that if we have the lower court names, we could create a list with lower court names, so that the court names are normalized

flooie commented on cl/scrapers/management/commands/cl_scrape_opinions.py:
```diff
@@ -103,6 +103,7 @@ def make_objects(
         item.get("source") or Docket.SCRAPER,
         blocked=blocked,
         date_blocked=date_blocked,
+        appeal_from_str=item.get("lower_court_str", ""),
```
why is this part of this PR

quevon24 replied:
Because we use that function to create the objects but currently does not allow saving lower court, that's why I had to update it

flooie commented on cl/opinion_page/forms.py:
```diff
@@ -401,45 +465,41 @@ def clean_pdf_upload(self) -> bytes:
         self.cleaned_data["item"] = {
             "source": Docket.DIRECT_INPUT,
             "cluster_source": SOURCES.DIRECT_COURT_INPUT,
+            "cluster_disposition": self.cleaned_data.get("disposition", ""),
```
the language here is inconsistent really with how we use it.  

I'm not sure why you are using cluster_disposition instead of just disposition 

same with date_filed.

quevon24 replied:
yes, i already checked carefully, i saw that the only model that has the disposition field is OpinionCluster, so it is only "disposition"

regarding date_filed, it has been defined as item["case_dates"] since 2019, making the change would imply modifying the code where the make_objects function is used, that could be done in an independent PR

flooie commented on cl/opinion_page/forms.py:
```diff
@@ -215,6 +222,20 @@ class CourtUploadForm(forms.Form):
+        widget=forms.TextInput(
+            attrs={
+                "class": "form-control",
+                "placeholder": "Judges / Lead author string, use this in case you can't "
```
I think you are conflating author/author_str on the opinion object with judges and panel on the cluster.  

```
    author = models.ForeignKey(
        "people_db.Person",
        help_text="The primary author of this opinion as a normalized field",
        related_name="opinions_written",
        on_delete=models.RESTRICT,
        blank=True,
        null=True,
    )
    author_str = models.TextField(
        help_text=(
            "The primary author of this opinion, as a simple text "
            "string. This field is used when normalized judges cannot "
            "be placed into the author field."
        ),
        blank=True,
    )
```

quevon24 replied:
So there should be a Charfield field in the form for judges and another for author?

quevon24 commented on cl/opinion_page/forms.py:
```diff
@@ -237,6 +258,70 @@ class CourtUploadForm(forms.Form):
+            }
+        ),
+    )
+
+    opinion_type = forms.CharField(
+        label="Opinion Type",
+        required=True,
+        initial=Opinion.LEAD,
+        widget=forms.Select(
+            choices=Opinion.OPINION_TYPES, attrs={"class": "form-control"}
+        ),
+    )
+
```
yes, you are right, i already checked the opinions and they are combined

quevon24 commented:
@flooie the changes are ready

- I refactored the form
- I removed the fields you suggested
- Now we use the seals from seal rookery (I am waiting for a response from the court to see if I can get the seal I found from Mississippi.) 
- I added some tests for the new forms and a test for maine
- Now we remove retired judges in tennessee form (#2515) 
- I added the judge_str field, and added the function to extract the last names
- I updated the make_objects function to save author_str in the opinion and removed the other changes

quevon24 commented:
> @quevon24, you flagged me as a reviewer, but if I'm needed I'd rather wait until @flooie has reviewed your second round of changes.
> 
> Can you flag me again then, please? (I just removed myself.)

Yes, no problem. I remembered that you commented on something here, that's why I also requested your review.

flooie commented on cl/opinion_page/forms.py:
```diff
@@ -250,141 +193,236 @@ def __init__(self, *args, **kwargs) -> None:
+        """
+        lead_author = forms.ModelChoiceField(
+            queryset=Person.objects.none(),
+            required=required,
```
This cant be required - at least not for all situations.  

we may not have many of the appellate judges in our system.

flooie commented on cl/opinion_page/forms.py:
```diff
@@ -158,81 +168,14 @@ class CourtUploadForm(forms.Form):
-    )
-    date_reargued = forms.DateField(
-        label="Reargued Date",
+    download_url = forms.URLField(
+        validators=[URLValidator()],
+        label="Download Url",
```
this field is now available in ME uploader - maybe tenn. but it wasn't /isnt in the current iteration.  can you make sure we dont alter the uploads for the other courts.

flooie commented on cl/opinion_page/forms.py:
```diff
@@ -250,141 +193,236 @@ def __init__(self, *args, **kwargs) -> None:
+            widget=forms.TextInput(
+                attrs={
+                    "class": "form-control",
+                    "placeholder": "Judges",
```
I think a better explanation of what judges is could be helpful.  

something similar to what the model says 

The judges that participated in the opinion as a simple "
            "text string. This field is used when normalized judges "
            "cannot be placed into the panel field.  not this exactly but something clearer.

quevon24 commented:
@flooie the changes are ready

- Added seal rookery as dependency
- Use seals for upload forms and court homepage
- Judges field placeholder changed
- Download url field is only available in new upload forms for miss and mo
- Lead author field is now not required
- I updated the view and court homepage template to use solr or elasticsearch based on the waffle flag (This is probably why you were not seeing the results correctly in your environment, I made the necessary adjustments)

flooie commented on cl/opinion_page/forms.py:
```diff
@@ -479,12 +509,299 @@ def save(self) -> OpinionCluster:
+                "cite_volume",
+                "cite_reporter",
+                "cite_page",
+                "download_url",
```
there is no download url right?

quevon24 replied:
right, I already removed it

flooie commented on cl/opinion_page/forms.py:
```diff
@@ -479,12 +509,299 @@ def save(self) -> OpinionCluster:
+                "cite_volume",
+                "cite_reporter",
+                "cite_page",
+                "download_url",
```
see above- no download url exists right?

quevon24 replied:
right, I already removed it

flooie commented on cl/opinion_page/views.py:
```diff
@@ -85,46 +90,89 @@
             results = f"results_{court}"
         else:
             results = "results"
-        response = await sync_to_async(do_search)(
-            request.GET.copy(),
-            override_params={
-                "filed_after": (
-                    datetime.datetime.today() - datetime.timedelta(days=28)
```
thanks for handling the elastic stuff but is there a reason we dropped sync to async for the not ES search

quevon24 replied:
I based it on another code in the repo, but async could be used, I already corrected it to use async

flooie commented on cl/opinion_page/views.py:
```diff
@@ -135,17 +183,40 @@ async def court_homepage(request: HttpRequest, pk: str) -> HttpResponse:
+        "me",
+        "mo",
+        "moctapp",
+        "moctapped",
```
if we are going to segment to moctapp ed wd and sd - we should probably not include a moctapp right?

quevon24 replied:
right, on the official site the opinions are only for each district

![image](https://github.com/freelawproject/courtlistener/assets/6474994/c2b5d15f-b52d-44fc-9130-d574dcea6fa1)

i already removed it

quevon24 commented:
The changes are ready @flooie 

- Added help text to fields
- Remove moctapp from views
- Remove unused download_url field from ordering
- Make court_homepage async again
- Fix merge conflict poetry.lock

flooie commented on cl/opinion_page/templates/court.html:
```diff
@@ -86,16 +107,34 @@ <h3>Recent Decisions of the&nbsp;{{ court }} <i class="gray fa fa-file-text-o"><
-      </div>
+      {% if pk == "me" %}
+        <div class="col-md-4">
+          <h3>About the Court</h3>
+          <p>The Supreme Judicial Court of Maine is the highest court in the
+            state
+            of Maine. It is made up of a chief justice and six associate
+            justices who are appointed by the governor and confirmed by the
+            Maine Senate. The court has general appellate jurisdiction and
+            is
+            responsible for interpreting the state's laws and constitution.
+            It
+            also has the power to issue advisory opinions on matters of law
+            to
+            the governor and the state's legislature.</p>
+          <hr>
+          <h3>Neutral Citations</h3>
+          <p>Supreme Judicial Court of Maine court uses neutral citations to
+            cite
+            its opinions.</p>
+          <p>Citations to the {{ court }} take the form:</p>
+          <pre>[YEAR] ME [#]</pre>
+          <p>In this cases, the format begins with the year, then an
+            abbreviation
+            for the court (ME), and finishes with a serial number indicating
+            the
+            decision number.</p>
+        </div>
```
some where whitespacing going on around here.

quevon24 replied:
i have already adjusted some of the white spaces to reduce the number of changes, there were several changes to the template and pycharm formatted it differently but i think it is because indent_size setting in .editorconfig file

flooie replied:
thanks for fixing that

flooie commented on cl/opinion_page/tests.py:
```diff
@@ -1240,6 +1318,64 @@ def test_form_save(self, mock) -> None:
+
+        self.assertEqual(pre_count + 1, Opinion.objects.all().count())
+
+    def test_mo_form_save(self, mock) -> None:
+        """Can we save missouri form successfully to db?"""
+
+        pre_count = Opinion.objects.all().count()
+
+        form = MoCourtUploadForm(
+            self.mo_data,
+            pk="mo",
+            files={"pdf_upload": self.pdf},
+        )
+
+        self.assertEqual(form.is_valid(), True, msg=form.errors)
+
+        if form.is_valid():
+            form.save()
+
+        self.assertEqual(pre_count + 1, Opinion.objects.all().count())
```
I think this test should measure if the opinion was successfully added to the right court - 

    def test_mo_form_save(self, mock) -> None:
        """Can we save missouri form successfully to db?"""

        pre_count = Opinion.objects.filter(cluster__docket__court__id="mo").count()
        form = MoCourtUploadForm(
            self.mo_data,
            pk="mo",
            files={"pdf_upload": self.pdf},
        )
        self.assertEqual(form.is_valid(), True, msg=form.errors)
        if form.is_valid():
            form.save()
        post_save_count = Opinion.objects.filter(cluster__docket__court__id="mo").count()
        self.assertEqual(pre_count + 1, post_save_count)

something more akin to this ... where we check that the opinion was added to the right court  no?

quevon24 replied:
yes,  this is a better approach, we validate that the opinion was added to the correct court, not just validate that we created an opinion

flooie commented on cl/opinion_page/tests.py:
```diff
@@ -1240,6 +1318,64 @@ def test_form_save(self, mock) -> None:
+
+        self.assertEqual(pre_count + 1, Opinion.objects.all().count())
+
+    def test_miss_form_save(self, mock) -> None:
+        """Can we save mississippi form successfully to db?"""
+
+        pre_count = Opinion.objects.all().count()
+
+        form = MissCourtUploadForm(
+            self.miss_data,
+            pk="miss",
+            files={"pdf_upload": self.pdf},
+        )
+
+        self.assertEqual(form.is_valid(), True, msg=form.errors)
+
         if form.is_valid():
```
same thing here

flooie commented on cl/opinion_page/tests.py:
```diff
@@ -1240,6 +1318,64 @@ def test_form_save(self, mock) -> None:
+
+        pre_count = Opinion.objects.all().count()
+
+        form = MeCourtUploadForm(
+            self.me_data,
+            pk="me",
+            files={"pdf_upload": self.pdf},
+        )
+
+        print("panel", form.fields["panel"].queryset)
```
can we drop the print here?

quevon24 replied:
yes

quevon24 commented on cl/scrapers/utils.py:
```diff
@@ -321,6 +323,7 @@ def update_or_create_docket(
         "date_blocked": date_blocked,
         "date_argued": date_argued,
         "ia_needs_upload": ia_needs_upload,
+        "appeal_from_str": appeal_from_str,
```
that's correct, I already removed it

flooie reviewed (commented):
This look good to me @quevon24 

Thanks for grinding thru on this.
